src/atomWithFormControls.ts

```diff
@@ -0,0 +1,187 @@
+  action: 'SET_VALUE' | 'SET_FOCUSED' | 'SET_TOUCHED';
+  key: string;
+  value: any;
+};
+
+const getDefaultOptions = <K extends symbol | string | number, V>() =>
+  <Options<K, V>>{
+    validate: (v) => v,
+  };
+
+export function atomWithFormControls<
+  AtomGroup extends Record<string, AtomWithValidation<any>>,
+  Keys extends keyof AtomGroup,
+  Vals extends AtomGroup[Keys],
+>(labeledAtoms: AtomGroup, options?: Options<Keys, Vals>) {
+  const { validate } = Object.assign(
+    {},
+    getDefaultOptions<Keys, Vals>(),
+    options,
+  );
+  const initBooleanState = Object.fromEntries(
+    Object.entries(labeledAtoms).map(([k]) => [k, false]),
+  );
+
+  const touchedState = atom(initBooleanState);
+  const focusedState = atom(initBooleanState);
+  const validating = validateAtoms(labeledAtoms, validate);
+
+  const errorsAtom = atom((get) => {
+    return Object.fromEntries(
+      Object.entries(labeledAtoms).map(([k, v]) => {
+        const val = get(v);
+        // @ts-expect-error atomgroup inference issue
+        if (val.isValid === false)
+          // @ts-expect-error result of the line above
+          return [k, val.error];
+        return [k, null];
+      }),
+    );
+  });
+
+  const valueAtom = atom(
+    (get) => {
+      return get(validating);
+    },
+    (get, set, next: ActionableNext) => {
+      switch (next.action) {
+        case 'SET_TOUCHED':
+          return set(touchedState, {
+            ...get(touchedState),
+            [next.key]: Boolean(next.value),
+          });
+        case 'SET_FOCUSED':
+          return set(focusedState, {
+            ...get(focusedState),
+            [next.key]: Boolean(next.value),
+          });
+        case 'SET_VALUE':
+        default: {
+          const actOn = labeledAtoms[next.key];
+          if (!actOn) {
+            return get(validating);
+          }
+          return set(actOn, next.value);
+        }
+      }
+    },
+  );
+
+  return atom(
+    (get) => {
+      const errorVals = get(errorsAtom);
+      const errLen = Object.keys(errorVals).filter((x) => errorVals[x]).length;
+      const validateAtomResult = get(valueAtom);
+      const isValid = validateAtomResult.isValid && errLen === 0;
+      return {
+        ...validateAtomResult,
+        isValid,
+        errors: errorVals,
+        touched: get(touchedState),
+        focused: get(focusedState),
+      };
+    },
+    (_, set, next: ActionableNext) => set(valueAtom, next),
+  );
+}
+
+type FormFieldValues<Keys extends string, Vals> = WritableAtom<
+  {
+    isValid: boolean | undefined;
+    errors: Record<Keys, any>;
+    touched: Record<Keys, boolean>;
+    focused: Record<Keys, boolean>;
+    values: Record<Keys, inferGeneric<Vals>>;
+    error: unknown;
+    isValidating: boolean | undefined;
+  },
+  [next: ActionableNext],
+  | void
+  | ({
+      values: Record<Keys, inferGeneric<Vals>>;
+    } & ValidatorState)
+>;
+
+export function useFormAtom<Keys extends string, Vals>(
```

Comment:
Up until now, this library is independent from React and it's a pure Jotai library. (It only depends on `jotai/vanilla`.)
This requires users to use React (or at least to include `react` package).
I think you have at least three options:
- Add `react` as a peer dependency and make this library React specific.
- Create a new entry point `jotai-form/react` and add `react` as an optional peer dependency.
- Don't include React specific functions in the code and just add it in docs.

Reply:
I'll be going with option 2 but then I'd still have to define a way the helper functions to exist in vanilla, I guess I'll just inject all the functions in the `atomWithFormControls`, I split them to avoid having a huge type definition but I forgot about jotai being JS first and react later